cabal-install/src/Distribution/Client/ProjectConfig.hs

```diff
@@ -747,8 +756,64 @@ readProjectLocalFreezeConfig verbosity httpTransport distDirLayout =
     "project freeze file"
 
 -- | Reads a named extended (with imports and conditionals) config file in the given project root dir, or returns empty.
-readProjectFileSkeleton :: Verbosity -> HttpTransport -> DistDirLayout -> String -> String -> Rebuild ProjectConfigSkeleton
+readProjectFileSkeleton :: HasCallStack => Verbosity -> HttpTransport -> DistDirLayout -> String -> String -> Rebuild ProjectConfigSkeleton
 readProjectFileSkeleton
+  verbosity
+  httpTransport
+  dir@DistDirLayout{distProjectFile, distDownloadSrcDirectory}
+  extensionName
+  extensionDescription = do
+    legacyPcs <- readProjectFileSkeletonLegacy verbosity httpTransport dir extensionName extensionDescription
+    exists <- liftIO $ doesFileExist extensionFile
+    if exists
+      then do
+        monitorFiles [monitorFileHashed extensionFile]
+        pcs <- liftIO $ readExtensionFile verbosity extensionFile
+        monitorFiles $ map monitorFileHashed (projectConfigPathRoot <$> projectSkeletonImports pcs)
+        unless (legacyPcs == pcs) (error (show callStack ++ "\nParsec: " ++ show pcs ++ "\nLegacy: " ++ show legacyPcs))
+        pure pcs
+      else do
+        monitorFiles [monitorNonExistentFile extensionFile]
+        return mempty
+    where
+      extensionFile = distProjectFile extensionName
+      readExtensionFile :: Verbosity -> FilePath -> IO ProjectConfigSkeleton
+      readExtensionFile verbosity' file = readAndParseFile (Parsec.parseProject extensionFile distDownloadSrcDirectory httpTransport verbosity' . ProjectConfigToParse) verbosity' file
+
+readAndParseFile
+  :: (BS.ByteString -> IO (Parsec.ParseResult a))
+  -> Verbosity
+  -> FilePath
+  -> IO a
+readAndParseFile parser verbosity fpath = do
+  exists <- doesFileExist fpath
```

Comment:
Do we care about the TOCTOU here?

Reply:
Yeah. Should we catch an exception from the readFile below instead?

Reply:
Is the new implementation alright? See https://github.com/haskell/cabal/pull/8889/files#diff-d5cf85f8d28570927985bfea598e5600b19afb53a759d65bb1c2172fc1dea3c5R845

When readFile throws, I catch the exception and throw a CabalException instead

Reply:
I think so, thanks! Let's see what Brandon has to say